core/trino-main/src/main/java/io/trino/sql/planner/iterative/rule/ImplementTableFunctionSource.java

```diff
@@ -0,0 +1,500 @@
+                    Optional.empty(),
+                    ImmutableSet.of(),
+                    0,
+                    Optional.empty(),
+                    node.getHandle()));
+        }
+
+        if (node.getSources().size() == 1) {
+            // Single source does not require pre-processing.
+            // If the source has row semantics, its specification is empty.
+            // If the source has set semantics, its specification is present, even if there is no partitioning or ordering specified.
+            // This property can be used later to choose optimal distribution.
+            TableArgumentProperties sourceProperties = getOnlyElement(node.getTableArgumentProperties());
+            return Result.ofPlanNode(new TableFunctionProcessingNode(
+                    node.getId(),
+                    node.getName(),
+                    node.getProperOutputs(),
+                    Optional.of(getOnlyElement(node.getSources())),
+                    ImmutableList.of(sourceProperties.getPassThroughSpecification()),
+                    ImmutableList.of(sourceProperties.getRequiredColumns()),
+                    Optional.empty(),
+                    sourceProperties.getSpecification(),
+                    ImmutableSet.of(),
+                    0,
+                    Optional.empty(),
+                    node.getHandle()));
+        }
+
+        if (node.getTableArgumentProperties().stream()
+                .anyMatch(TableArgumentProperties::isRowSemantics)) {
+            throw new UnsupportedOperationException("table arguments with row semantics are not yet supported when there are multiple table arguments"); // TODO
+        }
+        if (!node.getCopartitioningLists().isEmpty()) {
+            throw new UnsupportedOperationException("co-partitioned table arguments are not yet supported"); // TODO
+        }
+
+        // Collecting all partition size symbols to prune them out after preparing sources.
+        ImmutableList.Builder<Symbol> partitionSizeSymbols = ImmutableList.builder();
+
+        // Collecting all row number symbols for joined tables to prune them out after preparing sources.
+        ImmutableList.Builder<Symbol> combinedRowNumberSymbols = ImmutableList.builder();
+
+        // For all sources, map all source's output symbols to the source's row number symbol.
+        // The row number symbol will be later converted to a marker of "real" input rows vs "filler" input rows of the source.
+        // The "filler" input rows are the rows appended while joining partitions of different lengths,
+        // to fill the smaller partition up to the bigger partition's size. They are a side effect of the algorithm,
+        // and should not be processed by the table function.
+        ImmutableMap.Builder<Symbol, Symbol> rowNumberSymbolsBuilder = ImmutableMap.builder();
+
+        NodeWithSymbols first = appendHelperSymbolsForSource(0, node, partitionSizeSymbols, rowNumberSymbolsBuilder, metadata, context);
```

Comment:
What's a "helper symbol"?

Reply:
For each source we need to know the partition size and the row number within partition. They are necessary when making the cross-product of partitions from all sources. They are appended by a Window. "Helper symbols" applies to the row number symbol, and the partition size symbol.

Reply:
I would call the method "planWindowFunctions", instead. It's more descriptive of what's actually going on.